File: backend/teolgogo-api/src/main/java/com/teolgogo/entity/Payment.java
package com.teolgogo.entity;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import jakarta.persistence.*;
import java.time.LocalDateTime;

@Entity
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "payments")
public class Payment {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "customer_id", nullable = false)
    private User customer;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "business_id", nullable = false)
    private User business;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "quote_response_id", nullable = false)
    private QuoteResponse quoteResponse;

    private Integer amount; // 결제 금액

    @Enumerated(EnumType.STRING)
    private PaymentMethod paymentMethod; // 결제 수단

    @Enumerated(EnumType.STRING)
    private PaymentStatus status; // 결제 상태

    private String paymentKey; // PG사의 결제 키

    private String receiptUrl; // 영수증 URL

    private String orderId; // 주문 ID

    private LocalDateTime paidAt; // 결제 완료 시간

    private LocalDateTime createdAt; // 결제 요청 시간

    private LocalDateTime updatedAt; // 업데이트 시간

    @PrePersist
    public void prePersist() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        if (status == null) {
            status = PaymentStatus.PENDING;
        }
    }

    @PreUpdate
    public void preUpdate() {
        updatedAt = LocalDateTime.now();
    }

    // 결제 수단 enum
    public enum PaymentMethod {
        CARD, // 신용카드
        VIRTUAL_ACCOUNT, // 가상계좌
        ACCOUNT_TRANSFER, // 계좌이체
        PHONE, // 휴대폰 결제
        KAKAO_PAY, // 카카오페이
        TOSS_PAY, // 토스페이
        NAVER_PAY, // 네이버페이
        PAYCO // 페이코
    }

    // 결제 상태 enum
    public enum PaymentStatus {
        PENDING, // 결제 대기중
        READY, // 결제 준비됨
        IN_PROGRESS, // 결제 진행중
        DONE, // 결제 완료
        CANCELED, // 결제 취소
        FAILED, // 결제 실패
        EXPIRED // 결제 만료
    }
}

File: frontend/src/api/payment.ts
// import apiClient from './auth'; // 잘못된 import
import apiClient from './client'; // 올바른 API 클라이언트 import

// 결제 정보 인터페이스
export interface Payment {
  id: number;
  customerId: number;
  customerName?: string;
  businessId: number;
  businessName?: string;
  quoteResponseId: number;
  amount: number;
  paymentMethod: 'CARD' | 'VIRTUAL_ACCOUNT' | 'ACCOUNT_TRANSFER' | 'PHONE' | 'KAKAO_PAY' | 'TOSS_PAY' | 'NAVER_PAY' | 'PAYCO';
  status: 'PENDING' | 'READY' | 'IN_PROGRESS' | 'DONE' | 'CANCELED' | 'FAILED' | 'EXPIRED';
  paymentKey?: string;
  receiptUrl?: string;
  orderId: string;
  paidAt?: string;
  createdAt: string;
  updatedAt: string;
}

// 토스페이먼츠 결제 준비 응답
export interface TossPaymentPrepareResponse {
  paymentKey?: string;
  orderId: string;
  amount: number;
  customerKey?: string;
  successUrl: string;
  failUrl: string;
  windowTarget?: string;
  callbackUrl?: string;
  checkoutUrl: string; // 실제 결제 페이지 URL
  mobileRedirectUrl?: string;
}

// 카카오페이 결제 준비 응답
export interface KakaoPayPrepareResponse {
  tid: string; // 결제 고유 번호
  next_redirect_pc_url: string; // PC 웹 결제 페이지 URL
  next_redirect_mobile_url: string; // 모바일 웹 결제 페이지 URL
  next_redirect_app_url: string; // 모바일 앱 결제 페이지 URL
  created_at: string;
  android_app_scheme: string;
  ios_app_scheme: string;
  partner_order_id: string; // 주문 ID
}

// 토스페이먼츠 결제 준비
export const prepareTossPayment = async (quoteResponseId: number) => {
  try {
    const response = await apiClient.post('/payments/toss/prepare', null, {
      params: { quoteResponseId },
    });
    return response.data as TossPaymentPrepareResponse;
  } catch (error) {
    console.error('토스페이먼츠 결제 준비 실패:', error);
    throw error;
  }
};

// 카카오페이 결제 준비
export const prepareKakaoPayment = async (quoteResponseId: number) => {
  try {
    const response = await apiClient.post('/payments/kakao/prepare', null, {
      params: { quoteResponseId },
    });
    return response.data as KakaoPayPrepareResponse;
  } catch (error) {
    console.error('카카오페이 결제 준비 실패:', error);
    throw error;
  }
};

// 토스페이먼츠 결제 승인
export const confirmTossPayment = async (paymentKey: string, orderId: string, amount: number) => {
  try {
    const response = await apiClient.post('/payments/toss/confirm', null, {
      params: { paymentKey, orderId, amount },
    });
    return response.data;
  } catch (error) {
    console.error('토스페이먼츠 결제 승인 실패:', error);
    throw error;
  }
};

// 카카오페이 결제 승인
export const confirmKakaoPayment = async (
  pgToken: string,
  partnerOrderId: string,
  partnerUserId: string,
  tid: string
) => {
  try {
    const response = await apiClient.post('/payments/kakao/confirm', null, {
      params: { pg_token: pgToken, partner_order_id: partnerOrderId, partner_user_id: partnerUserId, tid },
    });
    return response.data;
  } catch (error) {
    console.error('카카오페이 결제 승인 실패:', error);
    throw error;
  }
};

// 결제 취소
export const cancelPayment = async (paymentId: number, cancelReason: string) => {
  try {
    const response = await apiClient.post(`/payments/${paymentId}/cancel`, null, {
      params: { cancelReason },
    });
    return response.data;
  } catch (error) {
    console.error('결제 취소 실패:', error);
    throw error;
  }
};

// 고객의 결제 내역 조회
export const getCustomerPaymentHistory = async () => {
  try {
    const response = await apiClient.get('/payments/customer/history');
    return response.data as Payment[];
  } catch (error) {
    console.error('결제 내역 조회 실패:', error);
    throw error;
  }
};

// 업체의 결제 내역 조회
export const getBusinessPaymentHistory = async () => {
  try {
    const response = await apiClient.get('/payments/business/history');
    return response.data as Payment[];
  } catch (error) {
    console.error('결제 내역 조회 실패:', error);
    throw error;
  }
};

// 결제 상세 조회
export const getPaymentDetails = async (paymentId: number) => {
  try {
    const response = await apiClient.get(`/payments/${paymentId}`);
    return response.data as Payment;
  } catch (error) {
    console.error('결제 상세 조회 실패:', error);
    throw error;
  }
};

// 견적 응답에 연결된 결제 정보 조회
export const getPaymentByQuoteResponse = async (quoteResponseId: number) => {
  try {
    const response = await apiClient.get(`/payments/by-quote-response/${quoteResponseId}`);
    return response.data as Payment;
  } catch (error) {
    console.error('결제 정보 조회 실패:', error);
    throw error;
  }
};

// 가상 결제 처리 (개발 및 테스트용)
export const processVirtualPayment = async (
  quoteResponseId: number,
  paymentMethod: Payment['paymentMethod'] = 'CARD'
) => {
  try {
    const response = await apiClient.post('/payments/virtual-payment', null, {
      params: { quoteResponseId, paymentMethod },
    });
    return response.data;
  } catch (error) {
    console.error('가상 결제 처리 실패:', error);
    throw error;
  }
};